### PR TITLE
Feature/sign up

### PR DESCRIPTION
**Pull request by elcorren**
## Summary
- User sign-up finished
- Migration file fixed

## Known Issues
None

## Trello Card
https://trello.com/c/BOllKerz/35-sign-up

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    const validPass = passwordLength(password);
+    if (!validPass) return res.json({ error: 'Invalid password.' });
+    const hashedPassword = await hashPassword(password);
+
+    // Creo el nuevo usuario
+    const newUser = await User.create({ email, password: hashedPassword, name, lastName });
+    if (!newUser) return res.json({ error: 'DB error.' });
+
+    return res.json({ email });
+  } catch (error) {
```
Comment (FrancoCoronel) [inline app/controllers/users.js:62]: Depending in how you handle error, this code probably will be change.
Follow the logic of this files

/app/errors.js
/app/middlewares/errors.js

File: app/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+// Hashea el password mediante bcrypt
+const hashPassword = async password => {
+  const salt = await bcrypt.genSaltSync(10);
+  const hashedPass = await bcrypt.hashSync(password, salt);
+  return hashedPass;
+};
+
+// Verifica la validez y el dominio del email
+const emailValidation = email => {
```
Comment (FrancoCoronel) [inline app/controllers/users.js:23]: What happen if the sended email is...

1. "        @wolox.com.ar"
2. "@wolox.com.ar      "
3. "    @wolox.com.ar    "

**Review by FrancoCoronel — CHANGES_REQUESTED** [review]

File: app/routes.js [hunk 3]
```diff
@@ -1,8 +1,10 @@
 // const controller = require('./controllers/controller');
 const { healthCheck } = require('./controllers/healthCheck');
+const { register } = require('./controllers/users');
 
 exports.init = app => {
   app.get('/health', healthCheck);
+  app.post('/users', register);
```
[the remark this answers is not in the document]
Comment (elcorren) [reply app/routes.js:7]: It is true! The name signUp is more transparent for this endpoint. But the first list item on the card is to implement /users endpoint.

File: app/controllers/users.js [hunk 4]
```diff
@@ -1,66 +1,43 @@
...
 exports.register = async (req, res, next) => {
-  const { email, password, name } = req.body;
-  const lastName = mapRequest(req.body);
+  const { email: dirtEmail, password, name, last_name: lastName } = req.body;
+  const email = dirtEmail.trim();
   try {
-    // Verifico que el request contenga todos los campos
     const userValid = email && password && name && lastName;
-    if (!userValid) res.json({ error: 'Missing fields.' });
+    if (!userValid) throw errors.createUserError('Missing fields.');
```
Comment (FrancoCoronel) [inline app/controllers/users.js:23]: createUserError is generic, I think that invalidParamsError is more specific. In this case, something went wrong with params. What do yo think??

File: app/controllers/users.js [hunk 5]
```diff
@@ -1,66 +1,43 @@
...
 
-    // Verifico que el mail sea válido
     const validEmail = emailValidation(email);
-    if (!validEmail) res.json({ error: 'Invalid email.' });
+    if (!validEmail) throw errors.createUserError('Invalid email.');
```
Comment (FrancoCoronel) [inline app/controllers/users.js:26]: same, email is not correct (invalid param error)

File: app/controllers/users.js [hunk 6]
```diff
@@ -1,66 +1,43 @@
...
+
+    const validPass = passwordLength(password);
+    if (!validPass) throw errors.createUserError('Invalid password.');
```
Comment (FrancoCoronel) [inline app/controllers/users.js:29]: same, password is invalid (Invalid param error)

File: app/controllers/users.js [hunk 7]
```diff
@@ -1,66 +1,43 @@
...
 
-    // Verifico que el email no exista previamente
     const emailExists = await User.findOne({ where: { email } });
-    if (emailExists) return res.send({ error: 'Email already exists.' });
+    if (emailExists) throw errors.createUserError('Email already exists.');
```
Comment (FrancoCoronel) [inline app/controllers/users.js:32]: In this case, the email exists... so here could be uniqueEmailError. Are you agree?

**Review by FrancoCoronel — CHANGES_REQUESTED** [review]

File: app/services/users.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+const { User } = require('../models');
+
+exports.emailExists = email => User.findOne({ where: { email } });
```
Comment (FrancoCoronel) [inline app/services/users.js:3]: You handle de databaseError here:
```.js
exports.verifyEmailUnique = email =>
  User.findOne({ where: { email } }).catch(error => {
    logger.error('Error while trying to get user by email', error.message);
    throw errors.databaseError(error.message);
  });
```

File: app/services/users.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+
+exports.create = (email, password, name, lastName) => User.create({ email, password, name, lastName });
```
Comment (FrancoCoronel) [inline app/services/users.js:5]: Its better if you catch de error here directly:
```.js
exports.createUser = userData =>
  User.create(userData).catch(error => {
    logger.error('Error while trying to create user', error.message);
    throw errors.databaseError(error.message);
  });
```

**Review by FrancoCoronel — CHANGES_REQUESTED** [review]

File: app/services/users.js [hunk 10]
```diff
@@ -1,5 +1,19 @@
 const { User } = require('../models');
+const logger = require('../logger');
+const errors = require('../errors');
 
-exports.emailExists = email => User.findOne({ where: { email } });
+exports.emailExists = email => {
```
Comment (FrancoCoronel) [inline app/services/users.js:5]: If you are returning the query response without mapping the result, you can write this function as:
```.js
exports.emailExists = email =>
  User.findOne({ where: { email } }).catch(error => {
    logger.error('Error while trying to get user by email', error.message);
    throw errors.databaseError(error.message);
  });
```

File: app/services/users.js [hunk 11]
```diff
@@ -1,5 +1,19 @@
...
+  const existsUser = User.findOne({ where: { email } }).catch(error => {
+    logger.error('Error while trying to get user by email', error.message);
+    throw errors.databaseError(error.message);
+  });
+  return existsUser;
+};
 
-exports.create = (email, password, name, lastName) => User.create({ email, password, name, lastName });
+exports.create = (email, password, name, lastName) => {
```
Comment (FrancoCoronel) [inline app/services/users.js:13]: idem than emailExists

**Review by FrancoCoronel — CHANGES_REQUESTED** [review]

**Review by FrancoCoronel — APPROVED** [review]
Great Job Tincho !!!

**Pull request merged**